File: dotnet/DesignPatterns/BehavioralDesignPatterns/ObserverDesignPattern/ObserverDesignPattern.Example2/WeatherStation.cs
namespace ObserverDesignPattern.Example2;

public class WeatherStation : ISubject
{

    private List<IObserver> observers;

    private float temperature;
    private float humidity;
    private float pressure;

    public WeatherStation()
    {
        observers = new List<IObserver>();
    }

    public void RegisterObserver(IObserver observer)
    {
        observers.Add(observer);
        Console.WriteLine($"New observer added. Total: {observers.Count}");
    }


    public void RemoveObserver(IObserver observer)
    {
        if (observers.Contains(observer))
        {
            observers.Remove(observer);
            Console.WriteLine($"Observer removed. Total: {observers.Count}");
        }
    }

    public void NotifyObservers()
    {
        foreach (var observer in observers)
        {
            observer.Update(temperature, humidity, pressure);
        }
    }

    public void SetMeasurements(float temperature, float humidity, float pressure)
    {
        this.temperature = temperature;
        this.humidity = humidity;
        this.pressure = pressure;

        MeasurementsChanged();
    }

    private void MeasurementsChanged()
    {
        NotifyObservers();
    }

    // Getters for the measurements (optional) 
    public float GetTemperature() => temperature;
    public float GetHumidity() => humidity;
    public float GetPressure() => pressure;
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/ObserverDesignPattern/ObserverDesignPattern.Example2/src/WeatherStation.ts
import { ISubject } from './ISubject';
import { IObserver } from './IObserver';

export class WeatherStation implements ISubject {
  private readonly _observers: IObserver[] = [];
  private _temperature: number = 0;
  private _humidity: number = 0;
  private _pressure: number = 0;

  registerObserver(observer: IObserver): void {
    this._observers.push(observer);
    console.log(`New observer added. Total: ${this._observers.length}`);
  }

  removeObserver(observer: IObserver): void {
    const index = this._observers.indexOf(observer);
    if (index !== -1) {
      this._observers.splice(index, 1);
      console.log(`Observer removed. Total: ${this._observers.length}`);
    }
  }

  notifyObservers(): void {
    for (const observer of this._observers) {
      observer.update(this._temperature, this._humidity, this._pressure);
    }
  }

  setMeasurements(temperature: number, humidity: number, pressure: number): void {
    this._temperature = temperature;
    this._humidity = humidity;
    this._pressure = pressure;
    this.notifyObservers();
  }

  getTemperature(): number { return this._temperature; }
  getHumidity(): number { return this._humidity; }
  getPressure(): number { return this._pressure; }
}
